### PR TITLE
release: allow creation of development releases

### LINKED ISSUE #62144: sg release create should fail if an internal release with the same version number exists
Right now, you can run multiple times `sg release create --version vX.Y.Z`  which is convenient when iterating. But there's a risk of accidentally overwriting existing images with the same version number, which would be very dangerous, as it would completely break the semantics around internal/promoted. You could end up promoting an in internal release that isn't what you thought would be. 

And because we're only registering the release in the _finalization_  phase of the internal release creation, we would fail *after* having overwritten the containers. 

Ultimately, because we have the SHA of the previous internal release we could with some efforts get back on our feets, that's really not something we want to be forced to do. 

Ideas for how to approach this: 

A/ We say that a given Z number is a special one reserved for testing (like `X.Y.666`). 

- We check if the release exists before anything else, and fail unless it's the special number. 
- While this sounds fun (being an Iron Maiden fan, this always makes me sing "Number of the beast" in my head) not everyone might enjoy the cultural reference (`777` might be better though). 
- A special number is tribal knowledge, even if we say something about it when creating the release, it's hard to discover. 
- If more than one person try to use the _test number_ we'll end with multiple jobs pushing different images at the same time. Nice way to have fun on a friday evening, right? 

Mentioned it because it's fun, but I don't think we should do this. 

B/ We introduce a `--release-development` flag where we mark the release as "release-development" on the release registry, and allow for creating it repeatedly (basically overwriting the previous one). 

- We check if the release exists before anything else, and fail if it does unless it's been tagged as development. 
- We block promotion of released tagged as "release-development". 

The question remains on how to iterate on the promotion phase, I wonder if we should pass in the steps the info that it's a release-development one, and thus we could swap the artifacts destinations to a private destination on the fly. 

I'm scratching a bit my head on this one, as the steps might become really verbose if we do so. But still, B/ would be fine without this, as we already have to take manual steps to do so.

### PR DESCRIPTION
**Pull request by BolajiOlajide**
Closes #62144

## Test plan

Manual testing. Will cut an internal release also.

### CODE REVIEW

**Comment by Chickensoupwithrice** [comment]
I think we're still missing "fail unless it's development" for overwriting releases, listed in https://github.com/sourcegraph/sourcegraph/issues/62144, but that belongs in releaseregistry since that one already blocks re-creating older versions [here](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/releaseregistry@a3abb0904989dd8805b5083fdf312e1d1968bc0f/-/blob/store/pgstore/store.go?L180-186)

**Review by Chickensoupwithrice — APPROVED** [review]
Looks good

**Comment by BolajiOlajide** [comment]
> I think we're still missing "fail unless it's development" for overwriting releases, listed in #62144, but that belongs in releaseregistry since that one already blocks re-creating older versions [here](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/releaseregistry@a3abb0904989dd8805b5083fdf312e1d1968bc0f/-/blob/store/pgstore/store.go?L180-186)
> 
> We'll also need to update the call in `release.yaml` which makes calls to releaseregistry.

Good catch. I'll add that.

File: release.yaml [hunk 1]
```diff
@@ -34,7 +34,8 @@ internal:
               "message": "Internal release build for {{version}}",
               "env": {
                     "RELEASE_INTERNAL": "true",
-                    "VERSION": "{{tag}}"
+                    "VERSION": "{{tag}}",
+                    "IS_DEVELOPMENT_RELEASE": "{{is_development}}"
```
Comment (jhchabran) [inline release.yaml:38]: Do you plan to patch the pipeline generator to skip the finalize phases in another PR?

**Review by jhchabran — APPROVED** [review]
little comments, but nothing preventing to merge

File: release.yaml [hunk 2]
```diff
@@ -153,6 +154,25 @@ test:
...
+            echo $body
+            echo "--- raw body ---"
+            exit $exit_code
+          fi
+
+          is_development=$(echo "$body" | jq -r '.is_development')
+          if [ "$is_development" = "true" ]; then
+            echo "cannot promote a development release"
+            exit 1
+          fi
```
Comment (BolajiOlajide) [inline release.yaml:175]: @jhchabran This ensures we don't promote development releases.

File: release.yaml (hunk 1 above)
Comment (BolajiOlajide) [reply release.yaml:38]: https://github.com/sourcegraph/sourcegraph/pull/62159/files#r1579748429

Comment (BolajiOlajide) [reply release.yaml:38]: My thinking is to not even let this get to the pipeline , that's why I added a step before the request to buildkite for promotion to check if the release is a dev release.

File: release.yaml [hunk 3]
```diff
@@ -153,6 +154,25 @@ test:
...
   create:
     steps:
+      - name: 'validate promotion criteria'
+        cmd: |
+          echo "validating promotion criteria"
+          body=$(wget --content-on-error -O- --header="Content-Type: application/json" "https://releaseregistry.sourcegraph.com/v1/releases/sourcegraph/{{version}}")
+          exit_code=$?
+
+          if [ $exit_code != 0 ]; then
+            echo "❌ Failed to promote release on release registry, got:"
```
Comment (jhchabran) [inline release.yaml:164]: nit: need to update the error message here

**Review by jhchabran — APPROVED** [review]

File: release.yaml (hunk 3 above)
Comment (BolajiOlajide) [reply release.yaml:164]: Acl.

**Pull request merged**